lib/image.ts

```diff
@@ -0,0 +1,118 @@
+import { runInNewContext } from 'vm';
+
+export class Image {
+	public static shimContent: Buffer;
+
+	public static async addLivepushShim(
+		buildContext: Readable,
+		dockerfileRegex = /^(?:\.\/)?Dockerfile$/,
+	): Promise<Readable> {
+		const extract = tar.extract();
+		const pack = tar.pack();
+
+		return new Promise<Readable>((resolve, reject) => {
+			extract.on('entry', async (header, stream, next) => {
+				const errHandle = (err: Error) => {
+					if (err != null) {
+						reject(err);
+					} else {
+						next();
+					}
+				};
+
+				if (dockerfileRegex.test(header.name)) {
+					// We need to process the dockerfile to add the COPY instruction for the shim
+					const dockerfileContent = Image.processDockerfile(
+						(await Image.streamToBuffer(stream)).toString(),
+					);
+					pack.entry(header, dockerfileContent, errHandle);
+				} else {
+					const entry = pack.entry(header, errHandle);
+					stream.pipe(entry);
+				}
+			});
+
+			extract.on('error', reject);
+			extract.on('finish', async () => {
+				// Add the shim to the build context
+				const shim = pack.entry(
+					{ name: '.balena/livepush.sh', size: Image.shimContent.length },
+					Image.shimContent,
+					err => {
+						if (err != null) {
+							reject(err);
+						} else {
+							pack.finalize();
+							resolve(pack);
+						}
+					},
+				);
+			});
+
+			buildContext.pipe(extract);
+		});
+	}
+
+	public static processDockerfile(dockerfileContent: string): string {
+		return dockerfileContent
+			.split(/\r?\n/)
+			.filter(l => l.length > 0)
+			.map(Image.replaceCmd)
+			.concat([
+				'COPY .balena/livepush.sh /bin/livepush.sh',
+				'RUN cp /bin/sh /bin/sh.real && cp /bin/livepush.sh /bin/sh',
+			])
+			.join('\n');
+	}
+
+	private static streamToBuffer(
+		stream: NodeJS.ReadableStream,
+	): Promise<Buffer> {
+		return new Promise<Buffer>((resolve, reject) => {
+			const chunks: Buffer[] = [];
+			stream.on('data', chunks.push.bind(chunks));
+			stream.on('error', reject);
+			stream.on('end', () => {
+				resolve(Buffer.concat(chunks));
+			});
+		});
+	}
+
+	public static replaceCmd(line: string): string {
+		const parts = line.trim().split(' ');
+		if (parts[0].toUpperCase() === 'CMD') {
+			const args = parts.slice(1).join(' ');
+			if (args.startsWith('[') && args.endsWith(']')) {
+				// Insert the shim into an array based CMD argument
+				// Use the JSON parser to read the CMD
+				try {
+					const arr = JSON.parse(args);
+					if (!_.isArray(arr)) {
+						throw new Error('Non-array argument passed');
+					}
+
+					return `CMD ["/bin/sh", "-c", "${arr
+						.map(a => JSON.stringify(a).slice(1, -1))
+						.join(' ')}"]`;
+				} catch (e) {
+					throw new DockerfileParseError(`Could not parse CMD command: ${e}`);
+				}
+			}
+			return line;
+		}
+
+		return line;
+	}
+}
+
+// Load the shim once
+(async () => {
+	Image.shimContent = await fs.readFile(require.resolve('./shims/livepush.sh'));
```

Comment:
With extremely slow IO, you could have a race condition here: `shimContent` being used before being set.

Reply:
Good point, fixed.